File: 20-actors/shionome/methods/test_grounding.py
"""test_grounding.py — 潮目 (shionome) stock-layer entity-grounding bridge. ADR-2606072200. Standalone.

Hermetic: the core functions are pure (fed inline fixtures), so the suite never depends on a
sibling actor's file contents. One conditional check exercises the real kabuto/hokorobi ledgers
IF present (and the fail-open path otherwise), so it is green either way.
"""
from __future__ import annotations

import pathlib

from _t import run
from grounding import (LAYER_GROUNDING, ground, ground_equities, grounding_roadmap,
                       hokorobi_institutions, kabuto_equity_constituents,
                       kanjo_disclosed_companies, load_ledger, systemic_overlay)

ROOT = pathlib.Path(__file__).resolve().parents[3]


# ── kabuto constituent extraction ────────────────────────────────────────────────
def _kab_fixture():
    return [
        {":company/id": "org.corp.us.apple", ":company/name": "Apple", ":company/ticker": "AAPL",
         ":company/country": "US", ":company/sector": ":electronics", ":company/market-cap-busd": 3300.0,
         ":company/sourcing": ":representative"},
        {":company/id": "org.corp.tw.tsmc", ":company/name": "TSMC", ":company/ticker": "2330.TW",
         ":company/market-cap-busd": 950.0, ":company/sourcing": ":representative"},
        {":company/id": "org.corp.jp.x", ":company/name": "Unsized Co", ":company/ticker": "9999.T"},  # no market-cap
        {":address/of": "org.corp.us.apple", ":address/city": "Cupertino"},   # NOT a company record
        {":supply.edge/from": "a", ":supply.edge/to": "b"},                   # NOT a company record
    ]


def test_kabuto_filters_non_company_records():
    cons = kabuto_equity_constituents(_kab_fixture())
    assert len(cons) == 3   # 3 companies; address + supply-edge skipped
    assert {c["name"] for c in cons} == {"Apple", "TSMC", "Unsized Co"}


def test_kabuto_parses_size_and_sector():
    cons = {c["name"]: c for c in kabuto_equity_constituents(_kab_fixture())}
    assert cons["Apple"]["market_cap_busd"] == 3300.0
    assert cons["Apple"]["sector"] == "electronics"     # keyword normalized
    assert cons["Unsized Co"]["market_cap_busd"] is None


# ── hokorobi institution extraction ──────────────────────────────────────────────
def _hok_fixture():
    return [
        {":organism/id": "fin.inst.jpmorgan", ":organism/kind": ":institution", ":organism/label": "JPMorgan Chase",
         ":inst/sector": ":bank", ":inst/sii": ":g-sib", ":inst/jurisdiction": "US", ":organism/sourcing": ":authoritative"},
        {":organism/id": "fin.inst.allianz", ":organism/kind": ":institution", ":organism/label": "Allianz",
         ":inst/sector": ":insurer", ":inst/sii": ":large", ":inst/jurisdiction": "DE", ":organism/sourcing": ":authoritative"},
        {":organism/id": "fin.inst.regional", ":organism/kind": ":institution", ":organism/label": "Regional bank",
         ":inst/sector": ":bank", ":inst/sii": ":mid", ":inst/jurisdiction": "US", ":organism/sourcing": ":representative"},
        {":organism/id": "risk.leverage", ":organism/kind": ":risk-source"},   # NOT an institution
        {":en/from": "a", ":en/to": "b", ":en/risk-load": 0.7},                # an edge, NOT an institution
    ]


def test_hokorobi_filters_to_institutions():
    insts = hokorobi_institutions(_hok_fixture())
    assert len(insts) == 3   # risk-source + edge skipped
    assert {i["label"] for i in insts} == {"JPMorgan Chase", "Allianz", "Regional bank"}


def test_systemic_overlay_counts_and_sourcing():
    ov = systemic_overlay(hokorobi_institutions(_hok_fixture()))
    assert ov["institutions"] == 3
    assert ov["by_sector"] == {"bank": 2, "insurer": 1}
    assert ov["authoritative"] == 2 and ov["representative"] == 1
    assert ov["no_trade_notice"] is True


# ── equities grounding math (the honesty contract) ───────────────────────────────
def test_ground_equities_value_coverage_math():
    cons = kabuto_equity_constituents(_kab_fixture())
    g = ground_equities(115.0, cons, universe=55000)
    assert g["grounded_entities"] == 3
    assert g["entities_with_size"] == 2
    # (3300 + 950) busd = 4.25 tn  →  4.25 / 115
    assert abs(g["grounded_market_cap_usd_tn"] - 4.25) < 1e-9
    assert abs(g["value_coverage_of_layer"] - round(4.25 / 115.0, 4)) < 1e-9


def test_ground_equities_value_coverage_is_lower_bound():
    # one sized + one unsized company → coverage is explicitly a lower bound
    g = ground_equities(115.0, kabuto_equity_constituents(_kab_fixture()))
    assert g["value_coverage_is_lower_bound"] is True


def test_ground_equities_count_coverage_fraction():
    g = ground_equities(115.0, kabuto_equity_constituents(_kab_fixture()), universe=55000)
    assert g["universe_sourcing"] == "representative"   # honesty: not a live count
    assert abs(g["count_coverage_of_universe"] - round(3 / 55000, 5)) < 1e-9


def test_ground_equities_top_constituents_sorted():
    g = ground_equities(115.0, kabuto_equity_constituents(_kab_fixture()))
    tops = g["top_constituents"]
    assert tops[0]["name"] == "Apple" and tops[1]["name"] == "TSMC"


def test_ground_equities_no_trade_notice():
    assert ground_equities(115.0, [])["no_trade_notice"] is True


def test_ground_equities_empty_is_zero_not_crash():
    g = ground_equities(115.0, [])
    assert g["grounded_entities"] == 0 and g["value_coverage_of_layer"] == 0.0


# ── full report + ungrounded honesty ─────────────────────────────────────────────
def _pyramid_fixture():
    return {"layers": [
        {"asset_class": "equities", "usd_tn": 115.0},
        {"asset_class": "gold", "usd_tn": 16.0},
        {"asset_class": "derivatives", "usd_tn": 600.0},
    ]}


def test_ground_full_report_shape():
    rep = ground(_pyramid_fixture(), _kab_fixture(), _hok_fixture())
    for k in ("equities", "systemic_institutions_overlay", "ungrounded_layers", "summary"):
        assert k in rep
    assert rep["summary"]["no_trade_notice"] is True


def test_ground_names_ungrounded_layers():
    rep = ground(_pyramid_fixture(), _kab_fixture(), _hok_fixture())
    # equities is grounded (kabuto present); gold + derivatives are NOT
    assert "equities" not in rep["ungrounded_layers"]
    assert set(rep["ungrounded_layers"]) == {"gold", "derivatives"}
    assert rep["summary"]["layers_with_entity_grounding"] == 1
    assert rep["summary"]["pyramid_layers"] == 3


def test_ground_total_named_entities():
    rep = ground(_pyramid_fixture(), _kab_fixture(), _hok_fixture())
    assert rep["summary"]["total_named_entities"] == 3 + 3   # 3 cos + 3 institutions


def test_no_kabuto_means_equities_ungrounded():
    # fail-open: no kabuto ledger → equities falls into ungrounded, no crash
    rep = ground(_pyramid_fixture(), [], _hok_fixture())
    assert "equities" in rep["ungrounded_layers"]
    assert rep["summary"]["layers_with_entity_grounding"] == 0


# ── kanjō disclosure depth (equities layer enrichment) ───────────────────────────
def _kanjo_fixture():
    return [
        {":fin.filing/id": "fil.a", ":fin.filing/company": "org.corp.us.apple", ":fin.filing/fiscal-year": 2024},
        {":fin.filing/id": "fil.t", ":fin.filing/company": "org.corp.tw.tsmc", ":fin.filing/fiscal-year": 2024},
        {":fin.fact/id": "fact.x", ":fin.fact/company": "org.corp.us.apple"},   # a fact, not a filing
    ]


def test_kanjo_disclosed_companies_extracts_filing_companies():
    disc = kanjo_disclosed_companies(_kanjo_fixture())
    assert disc == {"org.corp.us.apple", "org.corp.tw.tsmc"}   # facts ignored, dedup


def test_ground_equities_disclosure_depth():
    cons = kabuto_equity_constituents(_kab_fixture())   # ids: apple, tsmc, jp.x
    g = ground_equities(115.0, cons, disclosed_ids={"org.corp.us.apple", "org.corp.tw.tsmc"})
    assert g["with_disclosed_financials"] == 2
    assert set(g["disclosed_sample"]) == {"Apple", "TSMC"}


def test_ground_equities_depth_defaults_zero():
    g = ground_equities(115.0, kabuto_equity_constituents(_kab_fixture()))
    assert g["with_disclosed_financials"] == 0


# ── per-layer grounding roadmap ───────────────────────────────────────────────────
def test_roadmap_covers_every_layer():
    roadmap = grounding_roadmap(_pyramid_fixture())
    assert {r["asset_class"] for r in roadmap} == {"equities", "gold", "derivatives"}


def test_roadmap_equities_grounded_rest_ungroundable():
    roadmap = {r["asset_class"]: r for r in grounding_roadmap(_pyramid_fixture())}
    assert roadmap["equities"]["status"] == "grounded"
    assert roadmap["equities"]["source_actor"] == "kabuto"
    assert roadmap["gold"]["status"] == "ungroundable-at-r0"
    assert roadmap["derivatives"]["status"] == "ungroundable-at-r0"
    # every ungroundable layer states a non-empty reason (honesty contract)
    assert all(r["reason"] for r in roadmap.values())


def test_layer_grounding_registry_well_formed():
    for ac, spec in LAYER_GROUNDING.items():
        assert spec["status"] in ("grounded", "ungroundable-at-r0")
        assert "reason" in spec and spec["reason"]
        if spec["status"] == "grounded":
            assert spec["source_actor"]


def test_ground_report_has_roadmap_and_depth():
    rep = ground(_pyramid_fixture(), _kab_fixture(), _hok_fixture(), _kanjo_fixture())
    assert "roadmap" in rep and len(rep["roadmap"]) == 3
    assert rep["equities"]["with_disclosed_financials"] >= 1   # apple overlaps


def test_ground_without_kanjo_still_works():
    rep = ground(_pyramid_fixture(), _kab_fixture(), _hok_fixture())   # kanjo omitted
    assert rep["equities"]["with_disclosed_financials"] == 0


# ── fail-open loader ──────────────────────────────────────────────────────────────
def test_load_ledger_missing_returns_empty():
    assert load_ledger(ROOT / "20-actors" / "nonesuch" / "missing.edn") == []


def test_real_sibling_ledgers_present_or_fail_open():
    # exercises the REAL kabuto/hokorobi ledgers if checked out; green either way (fail-open).
    kab = load_ledger(ROOT / "20-actors" / "kabuto" / "data" / "seed-public-companies.kotoba.edn")
    if kab:
        assert len(kabuto_equity_constituents(kab)) > 100   # the real seed has >100 companies
    else:
        assert kabuto_equity_constituents(kab) == []        # fail-open path


if __name__ == "__main__":
    run("grounding", [(n, f) for n, f in sorted(globals().items())
                      if n.startswith("test_") and callable(f)])

File: 20-actors/shionome/methods/test_grounding.clj
#!/usr/bin/env bb
;; Clojure port of test_grounding.py — 潮目 (shionome) stock-layer entity-grounding bridge.
(ns shionome.methods.test-grounding
  "test_grounding.clj — 潮目 (shionome) stock-layer entity-grounding bridge. ADR-2606072200."
  (:require [clojure.test :refer [deftest is testing run-tests]]
            [shionome.methods.grounding :as g]))

(def ^:private ROOT
  (-> *file* java.io.File. .getAbsoluteFile .getParentFile
      .getParentFile .getParentFile .getParentFile))

;; ── kabuto constituent extraction ────────────────────────────────────────────────
(defn- kab-fixture []
  [{":company/id"              "org.corp.us.apple"
    ":company/name"            "Apple"
    ":company/ticker"          "AAPL"
    ":company/country"         "US"
    ":company/sector"          ":electronics"
    ":company/market-cap-busd" 3300.0
    ":company/sourcing"        ":representative"}
   {":company/id"              "org.corp.tw.tsmc"
    ":company/name"            "TSMC"
    ":company/ticker"          "2330.TW"
    ":company/market-cap-busd" 950.0
    ":company/sourcing"        ":representative"}
   {":company/id"     "org.corp.jp.x"
    ":company/name"   "Unsized Co"
    ":company/ticker" "9999.T"}                    ; no market-cap
   {":address/of" "org.corp.us.apple"
    ":address/city" "Cupertino"}                   ; NOT a company record
   {":supply.edge/from" "a" ":supply.edge/to" "b"} ; NOT a company record
   ])

(deftest test-kabuto-filters-non-company-records
  (let [cons (g/kabuto-equity-constituents (kab-fixture))]
    (is (= 3 (count cons)))   ; 3 companies; address + supply-edge skipped
    (is (= #{"Apple" "TSMC" "Unsized Co"} (set (map #(get % "name") cons))))))

(deftest test-kabuto-parses-size-and-sector
  (let [cons-map (into {} (map (fn [c] [(get c "name") c])
                               (g/kabuto-equity-constituents (kab-fixture))))]
    (is (= 3300.0 (get (get cons-map "Apple") "market_cap_busd")))
    (is (= "electronics" (get (get cons-map "Apple") "sector")))  ; keyword normalized
    (is (nil? (get (get cons-map "Unsized Co") "market_cap_busd")))))

;; ── hokorobi institution extraction ──────────────────────────────────────────────
(defn- hok-fixture []
  [{":organism/id"       "fin.inst.jpmorgan"
    ":organism/kind"     ":institution"
    ":organism/label"    "JPMorgan Chase"
    ":inst/sector"       ":bank"
    ":inst/sii"          ":g-sib"
    ":inst/jurisdiction" "US"
    ":organism/sourcing" ":authoritative"}
   {":organism/id"       "fin.inst.allianz"
    ":organism/kind"     ":institution"
    ":organism/label"    "Allianz"
    ":inst/sector"       ":insurer"
    ":inst/sii"          ":large"
    ":inst/jurisdiction" "DE"
    ":organism/sourcing" ":authoritative"}
   {":organism/id"       "fin.inst.regional"
    ":organism/kind"     ":institution"
    ":organism/label"    "Regional bank"
    ":inst/sector"       ":bank"
    ":inst/sii"          ":mid"
    ":inst/jurisdiction" "US"
    ":organism/sourcing" ":representative"}
   {":organism/id"   "risk.leverage"
    ":organism/kind" ":risk-source"}                ; NOT an institution
   {":en/from" "a" ":en/to" "b" ":en/risk-load" 0.7} ; an edge, NOT an institution
   ])

(deftest test-hokorobi-filters-to-institutions
  (let [insts (g/hokorobi-institutions (hok-fixture))]
    (is (= 3 (count insts)))  ; risk-source + edge skipped
    (is (= #{"JPMorgan Chase" "Allianz" "Regional bank"}
           (set (map #(get % "label") insts))))))

(deftest test-systemic-overlay-counts-and-sourcing
  (let [ov (g/systemic-overlay (g/hokorobi-institutions (hok-fixture)))]
    (is (= 3 (get ov "institutions")))
    (is (= {"bank" 2 "insurer" 1} (get ov "by_sector")))
    (is (= 2 (get ov "authoritative")))
    (is (= 1 (get ov "representative")))
    (is (true? (get ov "no_trade_notice")))))

;; ── equities grounding math (the honesty contract) ───────────────────────────────
(deftest test-ground-equities-value-coverage-math
  (let [cons (g/kabuto-equity-constituents (kab-fixture))
        ge   (g/ground-equities 115.0 cons 55000)]
    (is (= 3 (get ge "grounded_entities")))
    (is (= 2 (get ge "entities_with_size")))
    ;; (3300 + 950) busd = 4.25 tn → 4.25 / 115
    (is (< (Math/abs (- (get ge "grounded_market_cap_usd_tn") 4.25)) 1e-9))
    (is (< (Math/abs (- (get ge "value_coverage_of_layer")
                        (/ (Math/round (* (/ 4.25 115.0) 10000.0)) 10000.0)))
           1e-9))))

(deftest test-ground-equities-value-coverage-is-lower-bound
  ;; one sized + one unsized company → coverage is explicitly a lower bound
  (let [ge (g/ground-equities 115.0 (g/kabuto-equity-constituents (kab-fixture)))]
    (is (true? (get ge "value_coverage_is_lower_bound")))))

(deftest test-ground-equities-count-coverage-fraction
  (let [ge (g/ground-equities 115.0 (g/kabuto-equity-constituents (kab-fixture)) 55000)]
    (is (= "representative" (get ge "universe_sourcing")))  ; honesty: not a live count
    (is (< (Math/abs (- (get ge "count_coverage_of_universe")
                        (/ (Math/round (* (/ 3.0 55000.0) 100000.0)) 100000.0)))
           1e-9))))

(deftest test-ground-equities-top-constituents-sorted
  (let [ge   (g/ground-equities 115.0 (g/kabuto-equity-constituents (kab-fixture)))
        tops (get ge "top_constituents")]
    (is (= "Apple" (get (first tops) "name")))
    (is (= "TSMC"  (get (second tops) "name")))))

(deftest test-ground-equities-no-trade-notice
  (is (true? (get (g/ground-equities 115.0 []) "no_trade_notice"))))

(deftest test-ground-equities-empty-is-zero-not-crash
  (let [ge (g/ground-equities 115.0 [])]
    (is (= 0 (get ge "grounded_entities")))
    (is (= 0.0 (get ge "value_coverage_of_layer")))))

;; ── full report + ungrounded honesty ─────────────────────────────────────────────
(defn- pyramid-fixture []
  {"layers" [{"asset_class" "equities"    "usd_tn" 115.0}
             {"asset_class" "gold"        "usd_tn"  16.0}
             {"asset_class" "derivatives" "usd_tn" 600.0}]})

(deftest test-ground-full-report-shape
  (let [rep (g/ground (pyramid-fixture) (kab-fixture) (hok-fixture))]
    (doseq [k ["equities" "systemic_institutions_overlay" "ungrounded_layers" "summary"]]
      (is (contains? rep k)))
    (is (true? (get-in rep ["summary" "no_trade_notice"])))))

(deftest test-ground-names-ungrounded-layers
  (let [rep (g/ground (pyramid-fixture) (kab-fixture) (hok-fixture))]
    ;; equities is grounded (kabuto present); gold + derivatives are NOT
    (is (not (contains? (set (get rep "ungrounded_layers")) "equities")))
    (is (= #{"gold" "derivatives"} (set (get rep "ungrounded_layers"))))
    (is (= 1 (get-in rep ["summary" "layers_with_entity_grounding"])))
    (is (= 3 (get-in rep ["summary" "pyramid_layers"])))))

(deftest test-ground-total-named-entities
  (let [rep (g/ground (pyramid-fixture) (kab-fixture) (hok-fixture))]
    (is (= 6 (get-in rep ["summary" "total_named_entities"])))))  ; 3 cos + 3 institutions

(deftest test-no-kabuto-means-equities-ungrounded
  ;; fail-open: no kabuto ledger → equities falls into ungrounded, no crash
  (let [rep (g/ground (pyramid-fixture) [] (hok-fixture))]
    (is (contains? (set (get rep "ungrounded_layers")) "equities"))
    (is (= 0 (get-in rep ["summary" "layers_with_entity_grounding"])))))

;; ── kanjō disclosure depth (equities layer enrichment) ───────────────────────────
(defn- kanjo-fixture []
  [{":fin.filing/id"          "fil.a"
    ":fin.filing/company"     "org.corp.us.apple"
    ":fin.filing/fiscal-year" 2024}
   {":fin.filing/id"          "fil.t"
    ":fin.filing/company"     "org.corp.tw.tsmc"
    ":fin.filing/fiscal-year" 2024}
   {":fin.fact/id"      "fact.x"
    ":fin.fact/company" "org.corp.us.apple"}])  ; a fact, not a filing

(deftest test-kanjo-disclosed-companies-extracts-filing-companies
  (let [disc (g/kanjo-disclosed-companies (kanjo-fixture))]
    (is (= #{"org.corp.us.apple" "org.corp.tw.tsmc"} disc))))  ; facts ignored, dedup

(deftest test-ground-equities-disclosure-depth
  (let [cons (g/kabuto-equity-constituents (kab-fixture))  ; ids: apple, tsmc, jp.x
        ge   (g/ground-equities 115.0 cons g/LISTED_UNIVERSE
                                #{"org.corp.us.apple" "org.corp.tw.tsmc"})]
    (is (= 2 (get ge "with_disclosed_financials")))
    (is (= #{"Apple" "TSMC"} (set (get ge "disclosed_sample"))))))

(deftest test-ground-equities-depth-defaults-zero
  (let [ge (g/ground-equities 115.0 (g/kabuto-equity-constituents (kab-fixture)))]
    (is (= 0 (get ge "with_disclosed_financials")))))

;; ── per-layer grounding roadmap ───────────────────────────────────────────────────
(deftest test-roadmap-covers-every-layer
  (let [roadmap (g/grounding-roadmap (pyramid-fixture))]
    (is (= #{"equities" "gold" "derivatives"}
           (set (map #(get % "asset_class") roadmap))))))

(deftest test-roadmap-equities-grounded-rest-ungroundable
  (let [roadmap-m (into {} (map (fn [r] [(get r "asset_class") r])
                                (g/grounding-roadmap (pyramid-fixture))))]
    (is (= "grounded" (get (get roadmap-m "equities") "status")))
    (is (= "kabuto"   (get (get roadmap-m "equities") "source_actor")))
    (is (= "ungroundable-at-r0" (get (get roadmap-m "gold") "status")))
    (is (= "ungroundable-at-r0" (get (get roadmap-m "derivatives") "status")))
    ;; every ungroundable layer states a non-empty reason (honesty contract)
    (is (every? #(seq (get % "reason")) (vals roadmap-m)))))

(deftest test-layer-grounding-registry-well-formed
  (doseq [[ac spec] g/LAYER_GROUNDING]
    (is (contains? #{"grounded" "ungroundable-at-r0"} (get spec "status")))
    (is (seq (get spec "reason")))
    (when (= "grounded" (get spec "status"))
      (is (some? (get spec "source_actor"))))))

(deftest test-ground-report-has-roadmap-and-depth
  (let [rep (g/ground (pyramid-fixture) (kab-fixture) (hok-fixture) (kanjo-fixture))]
    (is (contains? rep "roadmap"))
    (is (= 3 (count (get rep "roadmap"))))
    (is (>= (get-in rep ["equities" "with_disclosed_financials"]) 1))))  ; apple overlaps

(deftest test-ground-without-kanjo-still-works
  (let [rep (g/ground (pyramid-fixture) (kab-fixture) (hok-fixture))]  ; kanjo omitted
    (is (= 0 (get-in rep ["equities" "with_disclosed_financials"])))))

;; ── fail-open loader ──────────────────────────────────────────────────────────────
(deftest test-load-ledger-missing-returns-empty
  (is (= [] (g/load-ledger (java.io.File. ROOT "20-actors/nonesuch/missing.edn")))))

(deftest test-real-sibling-ledgers-present-or-fail-open
  ;; exercises the REAL kabuto/hokorobi ledgers if checked out; green either way (fail-open).
  (let [kab (g/load-ledger (java.io.File. ROOT "20-actors/kabuto/data/seed-public-companies.kotoba.edn"))]
    (if (seq kab)
      (is (> (count (g/kabuto-equity-constituents kab)) 100))  ; the real seed has >100 companies
      (is (= [] (g/kabuto-equity-constituents kab))))))          ; fail-open path

(when (= *file* (System/getProperty "babashka.file"))
  (let [{:keys [fail error]} (clojure.test/run-tests 'shionome.methods.test-grounding)]
    (System/exit (if (zero? (+ fail error)) 0 1))))
